### PR TITLE
Reviewing Docs Part 1

### PR DESCRIPTION
**Pull request by jeff-a-chiang**
The following pages have been completed: 

-docs/try-it-out/index.rst
-choose-jenkins-pipeline-type.rst
-deploy-devops-tools.rst

### CODE REVIEW

File: try-it-out/docker/jenkins/resources/scripts/init_groovy/configure.groovy [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+// Create the Jenkins webhook within Sonarqube to communicate to Jenkins that analysis was completed
+
+String responseCode = "-1" 
+
+while (!responseCode.contains("204")){
+    try{
+
+    
+String webhook = "http://sdp-jenkins:8080/sonarqube-webhook/"
```
Comment (steven-terrana) [inline try-it-out/docker/jenkins/resources/scripts/init_groovy/configure.groovy]: fix indentation of this section please :)

File: try-it-out/docker/sonarqube/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# pecos-sonarqube
```
Comment (steven-terrana) [inline try-it-out/docker/sonarqube/README.md]: You can delete this file? doesn't seem relevant to the lab

File: try-it-out/docs/index.rst [hunk 3]
```diff
@@ -0,0 +1,56 @@
+.. _Try It Out: 
+
+Try It Out
+===========
+
+Welcome to the Try It Out page for the Solutions Delivery Platform!
```
Comment (steven-terrana) [inline try-it-out/docs/index.rst]: Try It Out *lab*.  we should try to be consistent in vernacular.

File: try-it-out/docs/index.rst [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+========
+Overview
+========
+
+The purpose of this guide is to help you set up the Solutions Delivery Platform on your own local computer so that you can try it out and see what magic it can do!
```
Comment (steven-terrana) [inline try-it-out/docs/index.rst:12]: This is probably a team discussion, but i usually prefer hard Line Wrapping.  I don't usually count.. but enough to see the entire thing on one screen without having to scroll.  

Should probably just standardize on IDE settings..

File: try-it-out/docs/index.rst [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+It will walk you though the steps of setting up core DevOps tools using Docker containers, creating a simple pipeline for some sample Spring Boot API code,
+and utilizing some of the key features of the Jenkins Templating Engine, such as :ref:`Organizational Governance <conditional inheritance>`.
+
+The pipeline itself will be able to build a Docker image from some Spring Boot API code and perform some code analysis on it as well using SonarQube.
```
Comment (steven-terrana) [inline try-it-out/docs/index.rst]: Watch out for unnecessary words :) 

lot's of "some"s here

File: try-it-out/docs/index.rst [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+Prerequisites
+=============
+
+    * A text editor
+    * You have `Git`_ installed on your machine 
+    * You have `Docker`_ installed on your machine
+
+        - For your Docker resources, you will need at least 2CPUs, 3.0GIB Memory, and 1.0GIB Swap to ensure that everything within this guide can run properly
+    
+        - Follow the sections under the Advanced section on these sites depending on your operating machine to change the resources you allocate to Docker: https://docs.docker.com/docker-for-windows/#advanced or https://docs.docker.com/docker-for-mac/#preferences-menu
```
Comment (steven-terrana) [inline try-it-out/docs/index.rst]: Embed those links :) 

makes it look more uniform

File: try-it-out/docs/index.rst [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+
+.. toctree::
+   :hidden:
+   :titlesonly:
+
+   pages/deploy-devops-tools
+   pages/choose-jenkins-pipeline-type
+   pages/create-repository-for-pipeline-config
+   configure_jenkins_in_chart
```
Comment (steven-terrana) [inline try-it-out/docs/index.rst]: It looks like you only need the lines start with pages/ 

get ridda the rest

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 8]
```diff
@@ -0,0 +1,60 @@
+.. _Deploy Devops Tools:
+-----------------------
+Deploy The Devops Tools
+-----------------------
+
+In this section, you will be deploying all of the DevOps tools that we will be using within this tutorial. This consists of deploying a Jenkins instance, a SonarQube server, and a local (insecure) Docker registry.
```
Comment (steven-terrana) [inline try-it-out/docs/pages/deploy-devops-tools.rst]: If this is really the first experience someone will have with SDP - the may not know what those tools are. 

maybe a ``csv-table`` that outlines the tool and a *brief* description of what it is?

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+In your terminal, navigate to the inside of the ``sdp-labs`` directory that you just cloned from GitHub.
+
+Run the following command in your terminal from the top-level directory of the ``sdp-labs`` directory to deploy the Docker containers using the `provided docker-compose.yaml file`_ within the ``try-it-out`` directory. For some more information about docker-compose files, visit `this page`_.
+
+.. _provided docker-compose.yaml file: https://github.com/boozallen/sdp-labs/blob/master/try-it-out/docker-compose.yaml
+
+.. code-block:: bash
+
+   docker-compose -f ./try-it-out/docker-compose.yaml up -d
```
Comment (steven-terrana) [inline try-it-out/docs/pages/deploy-devops-tools.rst]: does docker-compose up build images by default if they're not found?

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+.. _this page: https://docs.docker.com/compose/compose-file/
+
+The above command will create and run the following Docker containers: Jenkins, a Docker registry, and SonarQube. To verify this, you can run ``docker ps`` in your terminal. 
+
+The response should show something similar to the following screenshot.
+
+.. image:: ../images/deploy-devops-tools/docker_ps_command.png
+
+.. note:: If you looked at the ``docker-compose.yaml`` file we provided, you may have noticed that sonar-scanner isn't a running container. This is a Docker image that we will need later, even though it's intentionally not running.
```
Comment (steven-terrana) [inline try-it-out/docs/pages/deploy-devops-tools.rst]: too much information for a newbie.  keep it high level.

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+
+For your conenvience, we have already automated the following thing to allow you to focus on the usage of SDP:
```
Comment (steven-terrana) [inline try-it-out/docs/pages/deploy-devops-tools.rst]: It's the try it out.  with focus as a finite resource.. let's not give them more information than they actually need to get familiar with SDP. 

we automated it so that they don't have to think about this, so why even tell them we automated it.

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    
+    * Creation of Jenkins Credentials to be able to access SonarQube as well as your local Docker registry
+    * Downloading all the needed plugins required for the SDP including the Jenkins Templating Engine plugin
+    * Creation of a webhook within SonarQube to notify your Jenkins instance that code analysis was successfully completed
+
+=================================
+Register Insecure Docker Registry
+=================================
+
+To be able to use and access the Docker registry that you just deployed, you will need to add the Docker registry's IP Address and Port (which you can find using ``docker ps``) to the list of insecure registries found in your Docker Preferences.
```
Comment (steven-terrana) [inline try-it-out/docs/pages/deploy-devops-tools.rst:60]: If someone just installed docker and haven't used it before, they aren't going to know how to get a container's IP address. 

can't you just register ``localhost:5000`` as the insecure registry? 

if so - let's just directly tell them that. 
if not - let's get the command to get the container ip and give it to them.

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+Follow the instructions under the section labeled "Deploy a plain HTTP registry" on the following website to do so within the format of ``IP_Address:Port``: https://docs.docker.com/registry/insecure/#deploy-a-plain-http-registry
+
+.. note:: The URL of the Docker registry should not be preceded by **http://** when being added to the list of insecure registries.
```
Comment (steven-terrana) [inline try-it-out/docs/pages/deploy-devops-tools.rst]: At the bottom of this file add a 

========
Validate
========

section that walks through how to make sure everything got deployed correctly. 

i.e. "go to localhost:8080 and verify Jenkins is up and running", "go to localhost:9000 to validate sonarqube", etc

File: try-it-out/docs/pages/choose-jenkins-pipeline-type.rst [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+------------------------------
+Choose a Jenkins Pipeline Type
+------------------------------
+
+Now that you have all the services you need running on Docker, you will now be choosing the type of Jenkins pipeline that you will be configuring for a Spring Boot API sample application that we created just for this guide.
+
+To begin, navigate to the homepage of your Jenkins instance.
+
+Click **New Item** on the left-hand menu to create a new Jenkins pipeline.
```
Comment (steven-terrana) [inline try-it-out/docs/pages/choose-jenkins-pipeline-type.rst:11]: It would be better to embed a gif of you doing this.

File: try-it-out/docs/pages/choose-jenkins-pipeline-type.rst [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+Now that you have all the services you need running on Docker, you will now be choosing the type of Jenkins pipeline that you will be configuring for a Spring Boot API sample application that we created just for this guide.
+
+To begin, navigate to the homepage of your Jenkins instance.
+
+Click **New Item** on the left-hand menu to create a new Jenkins pipeline. 
+
+.. image:: ../images/choose-jenkins-pipeline-type/click_new_item.png
+
+For the **Item Name**, enter **sample-spring-boot-api**. Note that the **item name** you give it is arbitrary, but for consistency and referencing purposes, we'll be calling it **sample-spring-boot-api**. Below, you should see several options for the type of job to use. For the options that can currently be used with the Jenkins Templating Engine, they can be described as follows:
```
Comment (steven-terrana) [inline try-it-out/docs/pages/choose-jenkins-pipeline-type.rst]: I like having these docs somewhere, but i'm not positive this is the right place.  it's talking about concepts that haven't been introduced yet (pipeline configurations, etc)

File: try-it-out/docs/pages/choose-jenkins-pipeline-type.rst [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    
+    * Pipeline: Use this option when you want a one off job that may or may not be tied to a Source Control Management(SCM) tool.
+    * Folder: This option is useful for grouping things together that might not be connected otherwise to ensure they all inherit specific pipeline configurations. For example, if you had two GitHub repositories that were in different organizations, you could select this option, configure a configuration for it, and then create two new items (likely multipbranch pipelines) within the folder to be able to inherit from the original parent folder's pipeline configurations.
```
Comment (steven-terrana) [inline try-it-out/docs/pages/choose-jenkins-pipeline-type.rst]: This would be a great second lab.  moving up to a GitHub Organization job to shared configurations between repositories. 

and then another lab after that using a folder to add a final tier. 

the 'SDP Governance Hierarchy Lab Series' or something.

File: try-it-out/docs/pages/choose-jenkins-pipeline-type.rst [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    * GitHub Organization: This option is useful for setting a single pipeline configuration for all repositories under a specific GitHub Organization at the same time.
+    * Multibranch Pipeline: Use this option when you have a specific GitHub repository that you'd like to configure. When using this option, you cannot setup a child Git repository or Git organization to inherit from this item's configured pipeline configurations.
+
+For the purposes of this guide, we will be using a **Multibranch Pipeline**, as we will only be configuring a pipeline for a single GitHub repository. The settings for the new pipeline should look like the following screen:
+
+.. image:: ../images/choose-jenkins-pipeline-type/multibranch_pipeline_selected.png
+
+Click **OK**. You should be taken to a screen to configure your new pipeline. 
+
+To use the Jenkins Templating Engine, you will need to configure the configuration section
```
Comment (steven-terrana) [inline try-it-out/docs/pages/choose-jenkins-pipeline-type.rst:31]: This ordering doesn't make sense. 

You'll

for someone's first pipeline, you can just have them put a ``pipeline_config.groovy`` in their application repository. you dont actually **need** a pipeline configuration repository.

if you only have one application, it doesn't make sense anyway. 

the second lab around governance should introduce a pipeline configuration repository.

then this ordering issue goes away.

File: try-it-out/docs/pages/create-repository-for-pipeline-config.rst [hunk 18]
```diff
@@ -0,0 +1,132 @@
+.. _Create Git Repository for a Pipeline Configuration:
```
Comment (steven-terrana) [inline try-it-out/docs/pages/create-repository-for-pipeline-config.rst:1]: I like the idea of "try it out" just putting a ``pipeline_config.groovy`` in the sample app repo. 

it would be good to have a page that walks through what is it and what it does though.

**Review by steven-terrana — CHANGES_REQUESTED** [review]
Like the direction it's going. 

overall feedback:

- the less words the better in most cases :) 
- look into an animated GIF for simple things instead of paragraphs of instructions and a few screenshots
- for this lab - we don't need a pipeline configuration repository. you can just put a pipeline_config.groovy file in the sample app repository.

File: try-it-out/docs/pages/deploy-devops-tools.rst (hunk 9 above)
Comment (jeff-a-chiang) [reply try-it-out/docs/pages/deploy-devops-tools.rst]: Yes! It generally gives the following error: 

"WARNING: Image for service jenkins was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`."

If you make a change to how Jenkins should be built (i.e. the configure.groovy file), you will need to use docker-compose up --build to capture those changes in a new Docker image.

File: try-it-out/docs/pages/deploy-devops-tools.rst (hunk 12 above)
Comment (jeff-a-chiang) [reply try-it-out/docs/pages/deploy-devops-tools.rst:60]: My worry is if the user is using Docker Toolbox, which, if I remember correctly, uses 192.168.99.100 as its IP address, then it wouldn't work for them. Docker has officially deprecated Docker Toolbox though, so it's up to you whether we should support a tool that's been declared a legacy. Otherwise, we should just be able to see if it works with localhost.

File: try-it-out/docs/pages/deploy-devops-tools.rst (hunk 9 above)
Comment (jeff-a-chiang) [reply try-it-out/docs/pages/deploy-devops-tools.rst]: I've modified the command to be:

docker-compose -f ./try-it-out/docker-compose.yaml up --build -d --scale sonar-scanner=0

File: try-it-out/docs/pages/choose-jenkins-pipeline-type.rst [hunk 19]
```diff
@@ -0,0 +1,32 @@
+.. _Choose Jenkins Pipeline Type:
+
+------------------------------
+Choose a Jenkins Pipeline Type
+------------------------------
+
+Now that you have all the services you need running on Docker, you will now be choosing the type of Jenkins pipeline that you will be configuring for a Spring Boot API sample application that we created just for this guide.
+
+To begin, navigate to the homepage of your Jenkins instance.
```
Comment (kottoson) [inline try-it-out/docs/pages/choose-jenkins-pipeline-type.rst:9]: Maybe add a *(http://localhost:8080)*  link here?

File: try-it-out/docs/pages/choose-jenkins-pipeline-type.rst [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+For the purposes of this guide, we will be using a **Multibranch Pipeline**, as we will only be configuring a pipeline for a single GitHub repository. The settings for the new pipeline should look like the following screen:
+
+.. image:: ../images/choose-jenkins-pipeline-type/multibranch_pipeline_selected.png
+
+Click **OK**. You should be taken to a screen to configure your new pipeline. 
+
+To use the Jenkins Templating Engine, you will need to configure the configuration section 
+labeled Jenkins Templating Engine to specify the Git repository where your pipeline configurations are stored,
+the relative path to your Jenkinsfile, and the location of any Git repositories that contain library sources, which are essentially many of the tools that the Solutions Delivery Platform team has created to be already integrated with this platform (i.e. Git, Docker, SonarQube, Twistlock, etc.).
```
Comment (kottoson) [inline try-it-out/docs/pages/choose-jenkins-pipeline-type.rst:33]: What do you think about this wording?
“To use the JTE, you will need to fill out the section labeled Jenkins Templating Engine. Specify the Git repository where your pipeline configurations are stored, the relative path to your Jenkinsfile, and the location of any Git repositories that contain library sources. If using the SDP, one of those library sources should be |the sdp-libraries repository|, which contains curated libraries created by the Solutions Delivery Platform team for tools commonly used in CI/CD pipelines (Git, Docker, SonarQube, etc.).”

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 21]
```diff
@@ -0,0 +1,63 @@
+.. _Deploy Devops Tools:
+-----------------------
+Deploy The Devops Tools
+-----------------------
+
+In this section, you will be deploying all of the DevOps tools that we will be using within this tutorial. This consists of deploying a Jenkins instance, a SonarQube server, and a local (insecure) Docker registry.
```
Comment (kottoson) [inline try-it-out/docs/pages/deploy-devops-tools.rst]: What do you think of this wording?
 “In this section, you will be deploying all of the DevOps tools that we will be using within this tutorial: Jenkins, SonarQube, and a local (insecure) Docker registry.

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+.. _SDP-Labs: https://github.com/boozallen/sdp-labs
+
+.. code-block:: bash
+
+   ## cloning via ssh
+   git clone git@github.com:boozallen/sdp-labs.git
+   ## cloning via https
+   git clone https://github.com/boozallen/sdp-labs.git
+
+**If you haven't already**, use one of the above commands to clone the repository to your computer.
```
Comment (kottoson) [inline try-it-out/docs/pages/deploy-devops-tools.rst:34]: I know we used the "**If you haven't already**" disclaimer in previous docs b/c we had the instruction replicated in the docs. Since that's no longer true, is it reasonable to say this can be inferred?

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+    
+    * Creation of Jenkins Credentials to be able to access SonarQube as well as your local Docker registry
+    * Downloading all the needed plugins required for the SDP including the Jenkins Templating Engine plugin
+    * Creation of a webhook within SonarQube to notify your Jenkins instance that code analysis was successfully completed
+
+=================================
+Register Insecure Docker Registry
+=================================
+
+To be able to use and access the Docker registry that you just deployed, you will need to add the Docker registry's IP Address and Port (which you can find using ``docker ps``) to the list of insecure registries found in your Docker Preferences.
```
Comment (kottoson) [inline try-it-out/docs/pages/deploy-devops-tools.rst:60]: Does docker ps show the registry’s IP address? I anticipate questions about this paragraph

File: try-it-out/docs/pages/deploy-devops-tools.rst [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+
+Follow the instructions under the section labeled "Deploy a plain HTTP registry" on the following website to do so within the format of ``IP_Address:Port``: https://docs.docker.com/registry/insecure/#deploy-a-plain-http-registry
```
Comment (kottoson) [inline try-it-out/docs/pages/deploy-devops-tools.rst:62]: What do you think about this wording?
 “|Follow the instructions on deploying a plain HTTP registry found here.| When you add your registry, the format is “IP_Address:Port” Maybe add a screenshot at the end?
